shared-bindings/busdevice/I2CDevice.c

```diff
@@ -0,0 +1,269 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+// This file contains all of the Python API definitions for the
+// busio.I2C class.
+
+#include "shared-bindings/microcontroller/Pin.h"
+#include "shared-bindings/busdevice/I2CDevice.h"
+#include "shared-bindings/util.h"
+#include "shared-module/busdevice/I2CDevice.h"
+
+#include "lib/utils/buffer_helper.h"
+#include "lib/utils/context_manager_helpers.h"
+#include "py/runtime.h"
+#include "supervisor/shared/translate.h"
+
+
+//| class I2CDevice:
+//|     """
+//|     Represents a single I2C device and manages locking the bus and the device
+//|     address.
+//|     :param ~busio.I2C i2c: The I2C bus the device is on
+//|     :param int device_address: The 7 bit device address
+//|     :param bool probe: Probe for the device upon object creation, default is true
+//|     .. note:: This class is **NOT** built into CircuitPython. See
+//|       :ref:`here for install instructions <bus_device_installation>`.
+//|     Example:
+//|     .. code-block:: python
+//|         import busio
+//|         from board import *
+//|         from adafruit_bus_device.i2c_device import I2CDevice
+//|         with busio.I2C(SCL, SDA) as i2c:
+//|             device = I2CDevice(i2c, 0x70)
+//|             bytes_read = bytearray(4)
+//|             with device:
+//|                 device.readinto(bytes_read)
+//|             # A second transaction
+//|             with device:
+//|                 device.write(bytes_read)"""
+//|     ...
+//| 
+STATIC mp_obj_t busdevice_i2cdevice_make_new(const mp_obj_type_t *type, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    busdevice_i2cdevice_obj_t *self = m_new_obj(busdevice_i2cdevice_obj_t);
+    self->base.type = &busdevice_i2cdevice_type;
+    enum { ARG_i2c, ARG_device_address, ARG_probe };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_i2c, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_device_address, MP_ARG_REQUIRED | MP_ARG_INT },
+        { MP_QSTR_probe, MP_ARG_BOOL, {.u_bool = true} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    busio_i2c_obj_t* i2c = args[ARG_i2c].u_obj;
+
+    common_hal_busdevice_i2cdevice_construct(self, i2c, args[ARG_device_address].u_int, args[ARG_probe].u_bool);
+    if (args[ARG_probe].u_bool == true) {
+        common_hal_busdevice_i2cdevice___probe_for_device(self);
+    }
+
+    return (mp_obj_t)self;
+}
+
+STATIC mp_obj_t busdevice_i2cdevice_obj___enter__(mp_obj_t self_in) {
+    common_hal_busdevice_i2cdevice_lock(self_in);
```

Comment:
Use `MP_OBJ_TO_PTR()` when you pass an `mp_obj_t` down to a `common_hal` routine. In practice it makes no difference, but we might eventually have more strict type-checking.

Reply:
Fixed everywhere I called common_hal